### ISSUE TITLE
Creation of Endpoints for Cloud Functions causes circular dependency between ESP and Endpoint services

### ISSUE THREAD

**Issue by nikolay-bocharov** [hashicorp/terraform-provider-google#5528]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

```
$ terraform -v
Terraform v0.12.20
+ provider.google v3.5.0
```

### Affected Resource(s)

google_cloud_run_service
google_endpoints_service

### Terraform Configuration Files

```
resource "google_cloud_run_service" "run" {
  name = var.name
  location = var.location

  template {
    spec {
      containers {
        image = "gcr.io/endpoints-release/endpoints-runtime-serverless:2"
        env {
          name = "ENDPOINTS_SERVICE_NAME"
          value = google_endpoints_service.openapi_service.service_name
        }
      }
    }
  }
}

resource "google_endpoints_service" "openapi_service" {
  service_name = replace(google_cloud_run_service.run.status[0].url,  "https://", "")
  project = var.project_id
  openapi_config = var.config
}
```

### Expected Behavior
**Please note that I do not expect this deployment with circular dependency to work.**
But there must be a working scenario  for creating Endpoints for Google Function possibly followed different approach or with different API.
Current API seem not to permit such cases.

### Actual Behavior

The circular dependency error appears

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

Google Cloud endpoint documentation from here - https://cloud.google.com/endpoints/docs/openapi/get-started-cloud-functions reads the following steps to create Endpoint for Google Function

1. Create ESP service with Google Run  and remember the URL address allocated by Google for this service
2. Crete Endpoint service (Use that URL address obtained on previous step to build the endpoint openapi config)
3. Update the ESP service created on step 1 with the name of the Endpoint

But these steps make both resources dependent on each other and inevitably bring to circular dependency error in Terraform. 
There must be a working scenario  for creating Endpoints for Google Function. 

### References

- Google documentation for Endpoint for Google Function https://cloud.google.com/endpoints/docs/openapi/get-started-cloud-functions

**Comment by nikolay-bocharov** [comment]
I've found a workaround for this using [provisioners](https://www.terraform.io/docs/provisioners/index.html)

```
resource "google_endpoints_service" "openapi_service" {
  service_name = replace(var.espService.url,  "https://", "")
  project = var.project_id
  openapi_config = var.config
  provisioner "local-exec" {
    command = "gcloud beta run services update ${var.espService.name} --set-env-vars ENDPOINTS_SERVICE_NAME=${self.service_name} --project ${var.project_id} --platform=managed --region=${var.region}"
  }
}
```

**Comment by danawillow** [comment]
I think the best way to solve this would be to add a fine-grained resource for a single cloud run environment variable. That would also fix the issue of wanting to be able to say that some env vars exist, but not be authoritative about them, which IIRC has been brought up as a concern. @chrisst, does that align with some of the things you've been thinking about?

**Comment by chrisst** [comment]
Yes I have been thinking that the set up for the cloud run resource itself should be separated from the deploying of an actual revision. I think this scenario and those wanting to deploy from cloud build or other CI's but not have Terraform roll back a revision make it clear that managing both workflows in the same resource isn't ideal. Right now the API doesn't accept an "empty" request to set the cluster up without specifying the initial revision but I'll talk with the cloud run team and see if that's a viable thing to support. Otherwise we can possible set up a "fake" revision that could be overridden with a subsequent deploy.

**Comment by rbong** [comment]
It appears that setting `ENDPOINTS_SERVICE_NAME` is no longer the favored way of passing the OpenAPI config to ESPv2.
See [this error](https://github.com/GoogleCloudPlatform/esp-v2/blob/f8d98017f9bce27b9fe0a49bb93045558667ca08/docker/serverless/env_start_proxy.py#L22).

>If you wish to skip this step, please specify the name of the
 service in the ENDPOINTS_SERVICE_NAME environment variable.
 Note this deployment mode is **not** officially supported.

The recommended way of deploying is to customize and build the docker image with [this script](https://github.com/GoogleCloudPlatform/esp-v2/blob/master/docker/serverless/gcloud_build_image), and bake the OpenAPI config into the docker image, rather than having the image download the config from endpoints.

The new steps are linked in the [originally linked guide for using endpoints with functions](https://cloud.google.com/endpoints/docs/openapi/get-started-cloud-functions).

I think this thread is still relevant, because even using the new approach, there is still a circular dependency according to the docs - you have to set the "host" value inside of the docker config, which is pointing to cloud run itself. My approach is going to have similar problems - I am going to build my docker image to accept the config as an environment variable - so the solution of having a separate resource for google cloud run environment variables still solves the issue.

**Comment by rbong** [comment]
I've got something working for the new instructions, the new workaround follows the same pattern but is more complicated:

```terraform
resource "google_cloud_run_service" "api_function_proxy" {
  name     = "my-api-function-proxy"
  location = var.region

  template {
    spec {
      containers {
        # Default to the unmodified image on first deploy
        image = "gcr.io/endpoints-release/endpoints-runtime-serverless:2"

        env {
          name  = "ESPv2_ARGS"
          value = "--cors_preset=basic"
        }
      }
    }
  }

  lifecycle {
    # Image is set automatically by api_function_proxy_image resource
    ignore_changes = [
      template[0].spec[0].containers[0].image
    ]
  }
}

resource "null_resource" "api_function_proxy_image" {
  # Always trigger since we have no reliable way to check if the build process changed
  triggers = {
    time = timestamp()
  }

  provisioner "local-exec" {
    command = "make docker-deploy-espv2"

    environment = {
      GCP_PROJECT         = var.project
      GCP_REGION          = var.region
      ENDPOINTS_HOST      = google_endpoints_service.api.dns_address
      ENDPOINTS_CONFIG_ID = google_endpoints_service.api.config_id
      ESPV2_SERVICE_NAME  = google_cloud_run_service.api_function_proxy.name
    }
  }
}
```

Where `make docker-deploy-espv2` submits the docker image just like the [Google ESPv2 build script made for this purpose](https://github.com/GoogleCloudPlatform/esp-v2/blob/master/docker/serverless/gcloud_build_image) and also updates the cloud run image:

```bash
gcloud run deploy "${ESPV2_SERVICE_NAME}" \
  --project="${GCP_PROJECT}" \
  --platform=managed \
  --region="${GCP_REGION}" \
  --image="${API_ESPV2_IMAGE}"
```

You could probably make the trigger for the image a lot more intelligent, but I don't really want to risk not running it.

Definitely not ideal, the recommended approach now is arguably worse for Terraform because it can no longer be fixed by adding a new environment variable resource. Hopefully Google provides a better solution soon.

**Comment by neurolabs** [comment]
@rbong Thanks for sharing your approach. In that approach, can't we use google_endpoints_service.api.config_id as a trigger, because it will at least change when the config of the endpoints service changes? That won't cover updates to the base docker image, though.

**Comment by rbong** [comment]
@neurolabs that would work except for the docker image as you've noted.

**Comment by neurolabs** [comment]
Thanks, @rbong for the feedback.

In the hack section of my head, I am including another trigger part based on substr(timestamp()) that would force deployment every month ;).

**Comment by rbong** [comment]
@neurolabs my approach, which I didn't include because it's hacky and subjective, is to hash every file in the espv2 docker image build directory, use it for the tag name, and then check if the tag name exists before pushing it with gcloud build.

```bash
API_ESPV2_IMAGE="gcr.io/${GCP_PROJECT}/endpoints-runtime-serverless:${ENDPOINTS_CONFIG_ID}-`cat build/* | md5sum | cut -d ' ' -f 1`"

// ...

if gcloud container images describe "${API_ESPV2_IMAGE}" &> /dev/null; then
  echo "Image ${API_ESPV2_IMAGE} already exists, not rebuilding"
else
  echo "Submitting ESPv2 image"
  gcloud builds submit --tag "${API_ESPV2_IMAGE}" --project "${GCP_PROJECT}" build/
fi
```

The `gcloud run deploy` command is executed regardless of if the image exists or not since the Run image is ignored by Terraform.

**Comment by neurolabs** [comment]
I hopefully figured something out to rebuild on endpoint change and (base) image update only. Add an external data source that resolves the digest of the latest image:

```
data "external" "latest_endpoints_runtime_digest" {
  program = [
    "./scripts/latest-digest.sh",
    "gcr.io/endpoints-release/endpoints-runtime-serverless:2"
    ]
}

resource "null_resource" "api_function_proxy_image" {
  triggers = {
    api_config_id = google_endpoints_service.api.config_id
    image_digest = data.external.latest_endpoints_runtime_digest.result.digest
  }
...
}
```

scripts/latest-digest.sh looks like

```
#!/bin/bash

# Exit if any of the intermediate steps fail
set -e 

# fetch the latest digest for the given image
DIGEST=$(gcloud container images describe $1 --format="value(image_summary.digest)")

# Safely produce a JSON object containing the result value.
# jq will ensure that the value is properly quoted
# and escaped to produce a valid JSON string.
jq -n --arg digest "$DIGEST" '{"digest":$digest }'
```

What do you think @rbong ? It should be enough to rebuild on base image change AFAICS.
```

**Comment by rbong** [comment]
Should be enough to rebuild on base image change AFAIK

**Comment by Arconapalus** [comment]
I've been working on something similar. Instead of creating an new issue I'll apply right here. In my development I'm looking to accomplish passing in output from child module to parent module and then back to child module. Hind sight, cloud run module output to main.tf config output and then back to cloud endpoint module. Where cloud endpoint module will use the output of the parent module (main.tf) and use it as a var for openapi_spec.yml. 
here is my setup 
parent module 
    - main.tf
    - outputs.tf
    - variable.tf
child module
   - CloudRun
          - variable.tf
          - outputs.tf
          - cloudRun.tf
   - CloudEndpoints
          - variable.tf
          - outputs.tf
          - cloudRun.tf
```
# Main.tf
provider google {
    credentials = file(var.credentials_file)
    project = var.project_name
    region = var.region
    #zone = "us-central-a"
} 

# ------------------------------------------------------------------------------
#  cloud run
# ------------------------------------------------------------------------------

module "helloapp" {

   source = ".//modules/services/CloudRunSrv"
   name = "helloapp"
   #location = "us-central1"
   project = "project_name"
   dockerimg = "gcr.io/cloudrun/hello" 

}

module "CloudRunESP2"{
    source = ".//modules/services/CloudRunSrv"
    name = "apigateway"
    project = "project_name"

    dockerimg = "gcr.io/endpoints-release/endpoints-runtime-serverless:2"
}




# ------------------------------------------------------------------------------
#  variables for Cloud endpoints                                                       
# ------------------------------------------------------------------------------


module "cloudEndpoints" {
      
     source = ".//modules/services/cloudEndpoints"
     project = "project_name"
      
}

# ------------------------------------------------------------------------------
# cloud projects
# ------------------------------------------------------------------------------

# ------------------------------------------------------------------------------
# TF scripts to pass in vars
# ------------------------------------------------------------------------------


# data "template_file"
```

Output file of parent module (main.tf) 

```
output "helloapp" {
  value = module.helloapp.url
}


output "CloudRunESP2" {
  value = module.CloudRunESP2.urlesp
}
```

Module CloudEndpoints
```
data "template_file" "openapi_spec" {
  template = file("${path.module}/openapi_spec.yml", { CloudRunESPURL = output.CloudRunESP2.CloudRunESP2  , CloudStorageURL = output.helloapp.helloapp })
}


resource "google_endpoints_service" "api-service" {
  service_name   = "api.endpoints.project_name.cloud.goog"
  project        = var.project
  openapi_config    = data.template_file.openapi_spec.rendered
  

}

```

Is there a way to pass in output as variable for template_file?
I may be missing something here or overlooking someone in the docs. 


Up until now I have been redeploying manually after cloud run espv2 is configured. I created this in cloudbuild.yml for google to run the gcloud_build_image
```

# - name: gcr.io/cloud-builders/gcloud
#   entrypoint: 'bash'
#   args: 
#     - '-c'
#     - |-
#       chmod +x gcloud_build_image 
#       ./gcloud_build_image -s cloudesp-rsbrtp3nuq-uc.a.run.app -c 2020-06-19r0 -p $PROJECT_NAME

```
I would then make another module of cloud run and use the docker image after created.

**Comment by fredo838** [comment]
The only thing we really need to not deploy 2 images to Cloud Run - which is the big thing messing with terraform - is the HASH in the url: https://my-cloud-run-HASH-LOC.a.run.app. If we can compute this value, or retrieve it through an API before a new image is deployed, we can make this work. We would need a data source, `data.cloud_run_hash`, that has as output that HASH value.
The steps would be :
  1) terraform calculates the hash
  2) terraform deploys the endpoint with the `swagger.yaml` with the correct URL
  3) You build and push the image with the swagger included with local-exec (as google is really saying "run this bash script", there's not really any way around using local-exec)
  4) terraform creates the cloud run resource.

So now the problem is reduced to: how to get the HASH value? 
- Does anyone know about an API that yields this value (other then deploying a dummy image)? Mentioning @chrisst  as he seems to know something about the APIs.  
- Is it the HASH even unique per GCP project, or is it per Cloud Run deployment? 
- If these methods don't exist, what's the proper place to request this feature? 

Also a friendly reminder to anyone coming accross this issue to +1 the original post, as it indicates the importance of the issue to the developers working on this repository.

**Comment by Arconapalus** [comment]
> I've been working on something similar. Instead of creating an new issue I'll apply right here. In my development I'm looking to accomplish passing in output from child module to parent module and then back to child module. Hind sight, cloud run module output to main.tf config output and then back to cloud endpoint module. Where cloud endpoint module will use the output of the parent module (main.tf) and use it as a var for openapi_spec.yml.
> here is my setup
> parent module
> - main.tf
> - outputs.tf
> - variable.tf
> child module
> 
> * CloudRun
>   - variable.tf
>   - outputs.tf
>   - cloudRun.tf
> * CloudEndpoints
>   - variable.tf
>   - outputs.tf
>   - cloudRun.tf
> 
> ```
> # Main.tf
> provider google {
>     credentials = file(var.credentials_file)
>     project = var.project_name
>     region = var.region
>     #zone = "us-central-a"
> } 
> 
> # ------------------------------------------------------------------------------
> #  cloud run
> # ------------------------------------------------------------------------------
> 
> module "helloapp" {
> 
>    source = ".//modules/services/CloudRunSrv"
>    name = "helloapp"
>    #location = "us-central1"
>    project = "project_name"
>    dockerimg = "gcr.io/cloudrun/hello" 
> 
> }
> 
> module "CloudRunESP2"{
>     source = ".//modules/services/CloudRunSrv"
>     name = "apigateway"
>     project = "project_name"
> 
>     dockerimg = "gcr.io/endpoints-release/endpoints-runtime-serverless:2"
> }
> 
> 
> 
> 
> # ------------------------------------------------------------------------------
> #  variables for Cloud endpoints                                                       
> # ------------------------------------------------------------------------------
> 
> 
> module "cloudEndpoints" {
>       
>      source = ".//modules/services/cloudEndpoints"
>      project = "project_name"
>       
> }
> 
> # ------------------------------------------------------------------------------
> # cloud projects
> # ------------------------------------------------------------------------------
> 
> # ------------------------------------------------------------------------------
> # TF scripts to pass in vars
> # ------------------------------------------------------------------------------
> 
> 
> # data "template_file"
> ```
> 
> Output file of parent module (main.tf)
> 
> ```
> output "helloapp" {
>   value = module.helloapp.url
> }
> 
> 
> output "CloudRunESP2" {
>   value = module.CloudRunESP2.urlesp
> }
> ```
> 
> Module CloudEndpoints
> 
> ```
> data "template_file" "openapi_spec" {
>   template = file("${path.module}/openapi_spec.yml", { CloudRunESPURL = output.CloudRunESP2.CloudRunESP2  , CloudStorageURL = output.helloapp.helloapp })
> }
> 
> 
> resource "google_endpoints_service" "api-service" {
>   service_name   = "api.endpoints.project_name.cloud.goog"
>   project        = var.project
>   openapi_config    = data.template_file.openapi_spec.rendered
>   
> 
> }
> ```
> 
> Is there a way to pass in output as variable for template_file?
> I may be missing something here or overlooking someone in the docs.
> 
> Up until now I have been redeploying manually after cloud run espv2 is configured. I created this in cloudbuild.yml for google to run the gcloud_build_image
> 
> ```
> 
> # - name: gcr.io/cloud-builders/gcloud
> #   entrypoint: 'bash'
> #   args: 
> #     - '-c'
> #     - |-
> #       chmod +x gcloud_build_image 
> #       ./gcloud_build_image -s cloudesp-rsbrtp3nuq-uc.a.run.app -c 2020-06-19r0 -p $PROJECT_NAME
> ```
> 
> I would then make another module of cloud run and use the docker image after creation.

I Fixed this issue

**Comment by Arconapalus** [comment]
Food for thought, I notice that running TerraForm with TF_logg=debug that the hash of the cloud run is shown during the building process as url. I'm not quite familiar with API's (just diving in) but I've have seen this in Cloud Build when I ran the script to rebuild the new ESPv2 API gateway. Through my CI/CD pipeline is github>cloudbuild.app>terraform>GCPproject. 
The has is only unique to Cloud Run and is set to all Cloud Run services.

**Comment by Arconapalus** [comment]
The API value is status.url mention from a Google Cloud Developer. This value is retrievable after Cloud Run has be configured.

**Comment by fredo838** [comment]
@Arconapalus What do you mean "configured"? After the cloud run application is already deployed?

**Comment by Arconapalus** [comment]
@fredo838  yes that's correct, once the cloud run services has been deployed and created.

**Comment by Arconapalus** [comment]
I see that this fixed for modular-magician, I'm running into the same problem with TerraForm CLI in cloudbuild env.
Would this be the right place to make an issue?

**Comment by Arconapalus** [comment]
> I see that this fixed for modular-magician, I'm running into the same problem with TerraForm CLI in cloudbuild env.
> Would this be the right place to make an issue?

seems as this is fixed, just tested this with TerraForm 0.12.29 

Thank you

**Comment by neurolabs** [comment]
@Arconapalus  with which version of the google provider did you test?

**Comment by Arconapalus** [comment]
* provider.google: version = "~> 3.33"

**Comment by kurdybacha** [comment]
So what is your approach exactly @Arconapalus ? Would you mind sharing a sample here, please?

**Comment by Arconapalus** [comment]
@kurdybacha Are you talking about what I was developing or the API getting Cloud Run HASH?

**Comment by kurdybacha** [comment]
@Arconapalus thanks for the reply. 
For me real issue is that you have to make an ESPv2 image build after configuration change openapi.json as per official documentation. That means that you can't really manage ESPv2 configuration (e.g. openapi.json) and ESP Cloud Run deployment  from the same terraform file without extra hops as per examples above.
Original issue in this thread is about circular dependency between `google_endpoints_service` and `google_cloud_run_service` because of `ENDPOINTS_SERVICE_NAME` and that can be solved as per your example @Arconapalus with `api.endpoints.project_name.cloud.goog` and as summarized e.g. in this [blog post](https://www.sru.ie/post/google-endpoints-cloud-run-terraform/).

**Comment by danawillow** [comment]
This issue seems to be covering several different things now. I'm going to go ahead and close it since it sounds like at least some parts are resolved. If this is still an issue, please either file a new bug or comment on this one with an explanation of which parts have not yet been resolved.

**Comment by fredo838** [comment]
Nope, there's no step closer to a solution, and the circular dependency is still a problem. The problem is still the same: the cloud run documentation requires overwriting a previously deployed ESP container which introduces a circular dependency which terraform can't handle.

**Comment by danawillow** [comment]
Have you gotten a chance to read the posted solution at https://www.sru.ie/post/google-endpoints-cloud-run-terraform/? If so, what's missing that still needs to be fixed?

**Comment by rbong** [comment]
@danawillow have not tried the solution, however note that [the ENDPOINTS_SERVER_HOST method used in the link is no longer officially supported](https://github.com/hashicorp/terraform-provider-google/issues/5528#issuecomment-652153172). According to the Google Docs, you should rebuild the Docker image instead. Some people have provided examples for building this image above. This is not the same as the circular dependency issue but is worth noting.

**Comment by connor-ps** [comment]
Does anyone have any working solutions for this? I am still quite stuck :(

**Comment by connor-ps** [comment]
> Have you gotten a chance to read the posted solution at https://www.sru.ie/post/google-endpoints-cloud-run-terraform/? If so, what's missing that still needs to be fixed?

This post seems to be incorrect. The service endpoint  <service-name>.endpoints.<project-id>.cloud.goog  doesn't seem to resolve as the host portion of the cloud run service. Only the   * w.a.run.app  URL will resolve

**Comment by odashi** [comment]
I ran into this issue recently, and resolved it using custom domains similarly as just https://github.com/hashicorp/terraform-provider-google/issues/5528#issuecomment-953184386 mentioned.

Here is the cherry-picked code in my scripts.
Note that I didn't deploy a dummy Cloud Run service to fix the domain: deployed only the final ESPv2 instance after the Cloud Entpoints service has been deployed successfully.

```terraform
resource "google_endpoints_service" "openapi_service" {
  service_name   = var.endpoint_name
  project        = var.project
  openapi_config = ...config with the custom domain decided before configuring...
}

resource "null_resource" "openapi_proxy_image" {
  triggers = {...triggers related to my workflow...}

  # Script obtained from:
  # https://github.com/GoogleCloudPlatform/esp-v2/blob/master/docker/serverless/gcloud_build_image
  provisioner "local-exec" {
    command = <<EOS
      bash gcloud_build_image \
        -c ${google_endpoints_service.openapi_service.config_id} \
        -s ${var.endpoint_name} \
        -p ${var.project} \
        -v ${var.espv2_version}
    EOS
  }
}

resource "google_cloud_run_service" "endpoint" {
  template {
    spec {
      containers {
        image = format(
          "gcr.io/%s/endpoints-runtime-serverless:%s-%s-%s",
          var.project,
          var.espv2_version,
          var.endpoint_name,
          google_endpoints_service.openapi_service.config_id,
        )
        ...
      }
      ...
    }
    ...
  }
  ...
  depends_on = [null_resource.openapi_proxy_image]
}

resource "google_cloud_run_domain_mapping" "endpoint" {
  location = var.region
  name     = var.endpoint_name
  metadata {
    namespace = var.project
  }
  spec {
    route_name = google_cloud_run_service.endpoint.name
  }
}

resource "google_dns_record_set" "endpoint" {
  name         = "${var.endpoint_name}."
  managed_zone = var.dns_zone
  type         = "CNAME"
  ttl          = 300
  rrdatas      = ["ghs.googlehosted.com."]
}
```

**Comment by vyacheslav31** [comment]
> I ran into this issue recently, and resolved it using custom domains similarly as just [#5528 (comment)](https://github.com/hashicorp/terraform-provider-google/issues/5528#issuecomment-953184386) mentioned.
> 
> Here is the cherry-picked code in my scripts. Note that I didn't deploy a dummy Cloud Run service to fix the domain: deployed only the final ESPv2 instance after the Cloud Entpoints service has been deployed successfully.
> 
> ```terraform
> resource "google_endpoints_service" "openapi_service" {
>   service_name   = var.endpoint_name
>   project        = var.project
>   openapi_config = ...config with the custom domain decided before configuring...
> }
> 
> resource "null_resource" "openapi_proxy_image" {
>   triggers = {
>     config_id = google_endpoints_service.openapi_service.config_id
>   }
> 
>   # Script obtained from:
>   # https://github.com/GoogleCloudPlatform/esp-v2/blob/master/docker/serverless/gcloud_build_image
>   provisioner "local-exec" {
>     command = <<EOS
>       bash gcloud_build_image \
>         -c ${google_endpoints_service.openapi_service.config_id} \
>         -s ${var.endpoint_name} \
>         -p ${var.project} \
>         -v ${var.espv2_version}
>     EOS
>   }
> }
> 
> resource "google_cloud_run_service" "endpoint" {
>   template {
>     spec {
>       containers {
>         image = format(
>           "gcr.io/%s/endpoints-runtime-serverless:%s-%s-%s",
>           var.project,
>           var.espv2_version,
>           var.endpoint_name,
>           google_endpoints_service.openapi_service.config_id,
>         )
>         ...
>       }
>       ...
>     }
>     ...
>   }
>   ...
>   depends_on = [null_resource.openapi_proxy_image]
> }
> 
> resource "google_cloud_run_domain_mapping" "endpoint" {
>   location = var.region
>   name     = var.endpoint_name
>   metadata {
>     namespace = var.project
>   }
>   spec {
>     route_name = google_cloud_run_service.endpoint.name
>   }
> }
> 
> resource "google_dns_record_set" "endpoint" {
>   name         = "${var.endpoint_name}."
>   managed_zone = var.dns_zone
>   type         = "CNAME"
>   ttl          = 300
>   rrdatas      = ["ghs.googlehosted.com."]
> }
> ```

Beautiful, this is exactly what I was looking for. Thank you.

**Comment by yanweiguo** [comment]
Can we close this given https://github.com/hashicorp/terraform-provider-google/issues/5528#issuecomment-953184386 is the solution?

**Comment by tomnewton** [comment]
No you shouldn't. Google Cloud Run domain mapping isn't available in all regions, so this is still a pretty big issues for terraform users.
